### PR TITLE
Improve test cases

### PR DESCRIPTION
**Pull request by privat**
this cleanup a little the `submission` classes and related info and fix some related bugs

### CODE REVIEW

File: src/model/missions.nit [hunk 1]
```diff
@@ -109,6 +109,9 @@ end
 class TestCase
 	serialize
 
+	# The number of the test in the test-suite (starting with 1)
+	var number: Int
```
Comment (Morriar) [inline src/model/missions.nit:113]: I like this, the UI likes this, everybody like this.

File: src/model/engines/engine_base.nit [hunk 2]
```diff
@@ -45,6 +46,7 @@ class Engine
 		submission.compilation_failed = not ok
 		if not ok then
 			submission.status = "error"
+			submission.results.clear
```
Comment (Morriar) [inline src/model/engines/engine_base.nit]: I see factorizable code here.

File: src/model/engines/engine_base.nit [hunk 3]
```diff
@@ -34,6 +34,16 @@ class Engine
 	fun run(submission: Submission, config: AppConfig) do
 		submission.status = "pending"
 
+		# Not yet completed mission.
+		if submission.mission.testsuite.is_empty then
```
Comment (Morriar) [inline src/model/engines/engine_base.nit:38]: Good one

File: www/directives/missions/submit.html [hunk 4]
```diff
@@ -58,13 +58,13 @@ <h3 class='title text-warning'>Star progression</h3>
 		<event-panel event='event' ng-repeat='event in result.events.__items' />
 
 		<!-- Compilation error -->
-		<div class='card' ng-if='result.compilation_failed'>
```
Comment (Morriar) [inline www/directives/missions/submit.html:61]: In the same file, you should also replace `index` by `testRes.testcase.number`.

**Review by Morriar — COMMENTED** [review]
Mostly a +1, I would love to see a reroll with some factorization about submissions errors (same code used at least 3 times) and changes in the `submit.html` frontend to use the test case numbering.

File: www/directives/missions/submit.html (hunk 4 above)
Comment (privat) [reply www/directives/missions/submit.html:61]: Can do

File: src/model/engines/engine_base.nit (hunk 2 above)
Comment (privat) [reply src/model/engines/engine_base.nit]: Can do

**Comment by privat** [comment]
update: apply @Morriar's suggestions

File: src/model/engines/engine_base.nit [hunk 5]
```diff
@@ -36,30 +36,21 @@ class Engine
 
 		# Not yet completed mission.
 		if submission.mission.testsuite.is_empty then
-			submission.status = "error"
 			submission.compilation.title = "Work in progress"
 			submission.compilation.message = "There is no tests for the mission yet.\nPlease try again later."
```
Comment (Morriar) [inline src/model/engines/engine_base.nit]: Oups, didn't see that one one the first review:

> There is no test~~s~~

And while we are at it, may I suggest:

> There is no test for **this** mission yet.

File: src/model/engines/engine_base.nit [hunk 6]
```diff
@@ -80,6 +71,15 @@ class Engine
 		submission.update_status(config)
 	end
 
+	# Mark the `submission` as an error, and save it in the `config`
+	fun set_compilation_error(config: AppConfig, submission: Submission)
+	do
```
Comment (Morriar) [inline src/model/engines/engine_base.nit]: You see me in the obligation to refer to the [Holly Rules of the Repository, chapter 3, section 1, alinea 2](https://github.com/Morriar/Missions/blob/master/README.md#code-style).

I cite:

> **The do is on the same line as the function declaration.**

Please, just stop this non-sense... The war is over, you just don't want to admit it.

**Review by Morriar — CHANGES_REQUESTED** [review]

**Comment by Morriar** [comment]
I'm starting to like this whole `review` thing.

**Comment by privat** [comment]
apply more suggestions, again and again...

**Review by Morriar — APPROVED** [review]

**Pull request merged**